lib/charms/opensearch/v0/opensearch_nodes_exclusions.py

```diff
@@ -41,75 +42,190 @@ class OpenSearchExclusions:
+        # cluster anymore.
+        to_remove = []
+        for node in removable:
+            if node not in units:
+                # Unit still exists
+                to_remove.append(node)
+        return set(to_remove)
+
+    def _add_voting(self, exclusions: Optional[Set[str]] = None) -> bool:
         """Include the current node in the CMs voting exclusions list of nodes."""
         try:
-            self._opensearch.request(
+            to_add = exclusions or {self._node.name}
+            response = self._opensearch.request(
                 "POST",
-                f"/_cluster/voting_config_exclusions?node_names={self._node.name}&timeout=1m",
+                f"/_cluster/voting_config_exclusions?node_names={','.join(to_add)}&timeout=1m",
                 alt_hosts=self._charm.alt_hosts,
                 resp_status_code=True,
-                retries=3,
             )
-            return True
+            logger.debug("Added voting, response: %s", response)
+
+            self._charm.peers_data.put(self._scope, VOTING_TO_DELETE, ",".join(to_add))
+            # The voting excl. API returns a status only
+            return response >= 200 and response < 300
         except OpenSearchHttpError:
             return False
 
-    def _delete_voting(self) -> bool:
-        """Remove all the voting exclusions - cannot target 1 exclusion at a time."""
+    def _delete_voting(self, exclusions: Set[str]) -> bool:
+        """Remove all voting exclusions and then re-adds the subset that should stay.
+
+        The API does not allow to remove a subset of the voting exclusions, at once.
+        """
+        current_excl = self._fetch_voting()
+        logger.debug("Current voting exclusions: %s", current_excl)
+        if not current_excl:
+            to_stay = None
+        else:
+            to_stay = self._fetch_voting() - exclusions
+        if current_excl == to_stay:
+            # Nothing to do
+            logger.debug("No voting exclusions to delete, current set is %s", to_stay)
+            return True
+
         # "wait_for_removal" is VERY important, it removes all voting configs immediately
         # and allows any node to return to the voting config in the future
         try:
-            self._opensearch.request(
+            response = self._opensearch.request(
                 "DELETE",
                 "/_cluster/voting_config_exclusions?wait_for_removal=false",
                 alt_hosts=self._charm.alt_hosts,
                 resp_status_code=True,
             )
-            return True
+            if response < 200 or response >= 300:
+                logger.debug("Failed to remove voting exclusions, response %s", response)
+                return False
+
+            logger.debug("Removed voting for: %s", to_stay)
+            return to_stay and self._add_voting(to_stay)
         except OpenSearchHttpError:
-            return False
+            pass
+        return False
```

Comment:
what we had was cleaner I believe

Reply:
Ack, it is just a personal view: don't like a method without final return... Anyways, returned to the original setting!